Firestore/core/src/firebase/firestore/local/leveldb_transaction.h

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_LOCAL_LEVELDB_TRANSACTION_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_LOCAL_LEVELDB_TRANSACTION_H_
+
+#include <leveldb/db.h>
+#include <map>
+#include <memory>
+#include <set>
+#include <string>
+#include <utility>
+
+#if __OBJC__
+#if !defined(GPB_USE_PROTOBUF_FRAMEWORK_IMPORTS)
+#define GPB_USE_PROTOBUF_FRAMEWORK_IMPORTS 0
+#endif
+
+#if GPB_USE_PROTOBUF_FRAMEWORK_IMPORTS
+#import <Protobuf/GPBProtocolBuffers.h>
+#else
+#import "GPBProtocolBuffers.h"
+#endif
+#endif
+
+namespace firebase {
+namespace firestore {
+namespace local {
+
+typedef std::map<std::string, std::string> Mutations;
+typedef std::set<std::string> Deletions;
+
+/**
+ * LevelDBTransaction tracks pending changes to entries in leveldb, including
+ * deletions. It also provides an Iterator to traverse a merged view of pending
+ * changes and committed values.
+ */
+class LevelDBTransaction {
+ public:
+  LevelDBTransaction(std::shared_ptr<leveldb::DB> db,
+                     const leveldb::ReadOptions& readOptions,
```

Comment:
I had planned on doing something similar (although I did not know the magic static incantation) in a later PR, but this seems ok to do now. My major hesitation is that this is the third copy of `StandardReadOptions`, and we can't yet delete the rest of them. But, it's coming, so that's fine.